File: src/main/java/com/ccjiuhong/gui/swing/DownloaderSwing.java
package com.ccjiuhong.gui.swing;

import com.alibaba.fastjson.JSONObject;
import com.ccjiuhong.gui.swing.frame.MainFrame;
import com.ccjiuhong.gui.swing.frame.MainPanel;
import com.ccjiuhong.mgt.DefaultDownloadManager;
import com.ccjiuhong.mgt.DownloadManager;
import com.ccjiuhong.util.FileUtil;
import lombok.extern.slf4j.Slf4j;
import org.jfree.chart.ChartFactory;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.StandardChartTheme;
import org.jfree.data.time.TimeSeries;
import org.jfree.data.time.TimeSeriesCollection;
import org.jfree.data.xy.XYDataset;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import java.awt.*;
import java.io.File;

import static javax.swing.JOptionPane.WARNING_MESSAGE;

/**
 * Swing版本的下载器GUI
 *
 * @author G. Seinfeld
 * @since 2019/12/10
 */
@Slf4j
public class DownloaderSwing {
    // 读取配置文件
    public static File configFile = new File("config/config.json");
    public static JSONObject jsonConfig = JSONObject.parseObject(FileUtil.readText(configFile));
    // 读取语言包配置
    public static String lang = jsonConfig.getString("language");

    public static JSONObject languageConfig = JSONObject.parseObject(FileUtil.readText(new File("config/i18n/" + lang + ".json")));

    private MainFrame mainFrame;

    DownloaderSwing() {
        // 获取下载管理器的实例
        DownloadManager defaultDownloadManager = DefaultDownloadManager.getInstance();

        mainFrame = MainFrame.getInstance();

        // 创建JPanel作为容器，所有其他组件都添加到JPanel中
        MainPanel mainPanel = MainPanel.getInstance();
        mainFrame.add(mainPanel);


        // 菜单条
        JMenuBar jMenuBar = new JMenuBar();
        mainFrame.setJMenuBar(jMenuBar);

        JMenu jMenu = new JMenu(languageConfig.getString("task"));
        jMenuBar.add(jMenu);


        JFreeChart chart = createChart();
        // 添加新建下载任务的菜单项
        addNewTaskMenuItem(jMenu, defaultDownloadManager, chart);

        // 添加修改配置的菜单项
        addConfigCenterMenuItem(jMenu);

        mainFrame.setVisible(true);
    }

    private void addConfigCenterMenuItem(JMenu jMenu) {
        JMenuItem item1 = new JMenuItem(languageConfig.getString("configCenter"));
        jMenu.add(item1);
        item1.addActionListener(e -> {
            JDialog frame = new JDialog(mainFrame, languageConfig.getString("configCenter"), true);
            frame.setSize(600, 480);
            frame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);

            String s = FileUtil.readText(configFile);
            JTextArea configText = new JTextArea(s);
            configText.setFont(new Font("Microsoft YaHei", Font.PLAIN, 16));

            configText.getDocument().addDocumentListener(new DocumentListener() {
                @Override
                public void insertUpdate(DocumentEvent e) {
                    System.out.println("inserted");
                }

                @Override
                public void removeUpdate(DocumentEvent e) {
                    System.out.println("removed");
                }

                @Override
                public void changedUpdate(DocumentEvent e) {
                    System.out.println("changed");
                }
            });
            frame.getContentPane().add(configText);

            JMenuBar jMenuBar = new JMenuBar();
            frame.setJMenuBar(jMenuBar);

            JMenu menu = new JMenu("file");
            jMenuBar.add(menu);

            JMenuItem saveItem = new JMenuItem("save");
            menu.add(saveItem);
            saveItem.addActionListener(e1 -> {
                try {
                    JSONObject.parseObject(configText.getText());
                    FileUtil.writeText(configFile, configText.getText());
                    jsonConfig = JSONObject.parseObject(FileUtil.readText(configFile));

                    // 读取语言包配置
                    lang = jsonConfig.getString("language");
                    languageConfig = JSONObject.parseObject(FileUtil.readText(new File("config/i18n/" + lang + ".json")));
                    mainFrame.revalidate();
                } catch (Exception ex) {
                    log.error("当前内容不是正确的json，无法保存", ex);
                    JOptionPane.showMessageDialog(null, "当前内容不是正确的json，无法保存", "警告", WARNING_MESSAGE);
                }
            });

            frame.setVisible(true);
        });
    }

    private void addNewTaskMenuItem(JMenu jMenu, DownloadManager defaultDownloadManager, JFreeChart chart) {
        JMenuItem item0 = new JMenuItem(languageConfig.getString("newTask"));
        jMenu.add(item0);
        item0.addActionListener(e -> {
            String fileUrl = JOptionPane.showInputDialog(languageConfig.getString("inputAddress"));
            if (fileUrl != null) startMissionForUrl(fileUrl, defaultDownloadManager);
        });
    }

    /**
     * 创建数据集合
     *
     * @return dataSet
     */
    private XYDataset createDataSet() {
        // 实例化DefaultCategoryDataset对象
        TimeSeriesCollection dataSet = new TimeSeriesCollection();

        // 向数据集合中添加数据
        TimeSeries speedSeries = new TimeSeries(languageConfig.getString("time"));
        dataSet.addSeries(speedSeries);
        return dataSet;
    }

    /**
     * 创建JFreeChart对象
     *
     * @return chart
     */
    private JFreeChart createChart() {
        StandardChartTheme standardChartTheme = new StandardChartTheme("CN"); //创建主题样式
        standardChartTheme.setExtraLargeFont(new Font("隶书", Font.BOLD, 20)); //设置标题字体
        standardChartTheme.setRegularFont(new Font("宋体", Font.PLAIN, 15));//设置图例的字体
        standardChartTheme.setLargeFont(new Font("宋体", Font.PLAIN, 15)); //设置轴向的字体
        ChartFactory.setChartTheme(standardChartTheme);//设置主题样式
        // 通过ChartFactory创建JFreeChart
        return ChartFactory.createTimeSeriesChart(
                languageConfig.getString("speed"),
                languageConfig.getString("time"),
                "",
                createDataSet(),
                true,
                false,
                false);

    }

    /**
     * 开始任务
     *
     * @param fileUrl 文件下载地址
     */
    private void startMissionForUrl(String fileUrl, DownloadManager defaultDownloadManager) {
        String targetFileName = fileUrl.substring(fileUrl.lastIndexOf("/") + 1);
        int missionId = defaultDownloadManager.addMission(fileUrl, jsonConfig.getString("downloadDirectory"), targetFileName);
        defaultDownloadManager.startOrResumeMission(missionId);


    }

    public static void main(String[] args) {
        try {
            UIManager.setLookAndFeel("com.sun.java.swing.plaf.gtk.GTKLookAndFeel");
            new DownloaderSwing();
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | UnsupportedLookAndFeelException e) {
            log.error(e.getMessage(), e);
        }

    }
}

File: src/main/java/com/ccjiuhong/gui/swing/DownloaderSwing.kt
package com.ccjiuhong.gui.swing


import com.alibaba.fastjson.JSONObject
import com.ccjiuhong.download.DownloadManager
import com.ccjiuhong.gui.common.Configuration.*
import org.apache.commons.lang3.concurrent.BasicThreadFactory
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartPanel
import org.jfree.chart.JFreeChart
import org.jfree.chart.StandardChartTheme
import org.jfree.chart.plot.XYPlot
import org.jfree.data.time.Second
import org.jfree.data.time.TimeSeries
import org.jfree.data.time.TimeSeriesCollection
import org.jfree.data.xy.XYDataset
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.Font
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.HeadlessException
import java.io.File
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit
import javax.swing.*
import javax.swing.JOptionPane.WARNING_MESSAGE
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener


/**
 * Swing版本的下载器GUI
 *
 * @author G.Seinfeld
 * @date 2019/06/30
 */
// 读取配置文件
val configFile = File("config/config.json")
var jsonConfig: JSONObject = JSONObject.parseObject(configFile.readText())

// 读取语言包配置
var lang = jsonConfig["language"]
var languageConfig: JSONObject = JSONObject.parseObject(File("config/i18n/$lang.json").readText())

class DownloaderSwing @Throws(HeadlessException::class)
constructor() : JFrame() {
    init {


        // 设置标题
        title = TITLE

        // 设置尺寸
        setSize(WINDOW_WIDTH, WINDOW_HEIGHT)

        // 在屏幕中央显示
        setLocationRelativeTo(null)

        // 窗口关闭后自动停止程序
        defaultCloseOperation = EXIT_ON_CLOSE

        // 获取下载管理器的实例
        val downloadManager = DownloadManager.getInstance()


        // 创建JPanel作为容器，所有其他组件都添加到JPanel中
        val gb = GridBagLayout()

        val jPanel = JPanel(gb)
        add(jPanel)


        val leftPanel = JPanel()
        var c = GridBagConstraints()
        c.fill = GridBagConstraints.BOTH
        c.weightx = 0.3
        c.weighty = 1.0
        gb.setConstraints(leftPanel, c)
        jPanel.add(leftPanel)


        val chart = createChart()

        val rightPanel = ChartPanel(chart)
        c = GridBagConstraints()
        c.fill = GridBagConstraints.BOTH
        c.weightx = 0.7
        c.weighty = 1.0
        gb.setConstraints(rightPanel, c)
        jPanel.add(rightPanel)


        // 菜单条
        val jMenuBar = JMenuBar()
        setJMenuBar(jMenuBar)

        val jMenu = JMenu(languageConfig["task"] as String?)
        jMenuBar.add(jMenu)

        // 添加新建下载任务的菜单项
        addNewTaskMenuItem(jMenu, downloadManager, chart)

        // 添加修改配置的菜单项
        addConfigCenterMenuItem(jMenu)

        isVisible = true
    }

    private fun addConfigCenterMenuItem(jMenu: JMenu) {
        val item1 = JMenuItem(languageConfig["configCenter"] as String?)
        jMenu.add(item1)
        item1.addActionListener {
            val s = configFile.readText()
            val frame = JDialog(this, languageConfig["configCenter"] as String?, true)
            frame.setSize(600, 480)
            frame.defaultCloseOperation = DISPOSE_ON_CLOSE

            val configText = JTextArea(s)
            configText.font = Font("Microsoft YaHei", Font.PLAIN, 16)
            configText.document.addDocumentListener(object : DocumentListener {
                override fun changedUpdate(e: DocumentEvent?) {
                    println("changed")
                }

                override fun insertUpdate(e: DocumentEvent?) {
                    println("inserted")
                }

                override fun removeUpdate(e: DocumentEvent?) {
                    println("removed")
                }

            })

            frame.contentPane.add(configText)

            val jMenuBar = JMenuBar()
            frame.jMenuBar = jMenuBar

            val menu = JMenu("file")
            jMenuBar.add(menu)

            val saveItem = JMenuItem("save")
            menu.add(saveItem)
            saveItem.addActionListener {
                try {
                    JSONObject.parseObject(configText.text)
                    configFile.writeText(configText.text)
                    jsonConfig = JSONObject.parseObject(configFile.readText())

                    // 读取语言包配置
                    lang = jsonConfig["language"]
                    languageConfig= JSONObject.parseObject(File("config/i18n/$lang.json").readText())
                    revalidate()
                    //repaint()
                } catch (e: Exception) {
                    logger.error("当前内容不是正确的json，无法保存", e)
                    JOptionPane.showMessageDialog(null, "当前内容不是正确的json，无法保存", "警告", WARNING_MESSAGE);
                }
            }

            frame.isVisible = true

        }
    }

    private fun addNewTaskMenuItem(jMenu: JMenu, downloadManager: DownloadManager, chart: JFreeChart) {
        val item0 = JMenuItem(languageConfig["newTask"] as String?)
        jMenu.add(item0)
        item0.addActionListener {
            val fileUrl = JOptionPane.showInputDialog(languageConfig["inputAddress"] as String?)
            if (fileUrl != null) startMissionForUrl(fileUrl, downloadManager, chart)
        }
    }

    /**
     * 创建数据集合
     * @return dataSet
     */
    private fun createDataSet(): XYDataset {
        // 实例化DefaultCategoryDataset对象
        val dataSet = TimeSeriesCollection()

        // 向数据集合中添加数据
        val speedSeries = TimeSeries(languageConfig["time"] as String?)
        dataSet.addSeries(speedSeries)
        return dataSet
    }

    /**
     * 创建JFreeChart对象
     * @return chart
     */
    private fun createChart(): JFreeChart {
        val standardChartTheme = StandardChartTheme("CN") //创建主题样式
        standardChartTheme.extraLargeFont = Font("隶书", Font.BOLD, 20) //设置标题字体
        standardChartTheme.regularFont = Font("宋体", Font.PLAIN, 15) //设置图例的字体
        standardChartTheme.largeFont = Font("宋体", Font.PLAIN, 15) //设置轴向的字体
        ChartFactory.setChartTheme(standardChartTheme)//设置主题样式
        // 通过ChartFactory创建JFreeChart
        return ChartFactory.createTimeSeriesChart(
                languageConfig["speed"] as String?,
                languageConfig["time"] as String?,
                "",
                createDataSet(),
                true,
                false,
                false)

    }

    /**
     * 开始任务
     * @param fileUrl 文件下载地址
     */
    private fun startMissionForUrl(fileUrl: String, downloadManager: DownloadManager, chart: JFreeChart) {
        val targetFileName = fileUrl.substring(fileUrl.lastIndexOf("/") + 1)
        val missionId = downloadManager.addMission(fileUrl, "F:\\rocketDownloader", targetFileName)
        downloadManager.startMission(missionId)

        var second = Second()
        val series = ((chart.plot as XYPlot).dataset as TimeSeriesCollection).series[0] as TimeSeries

        val executorService = ScheduledThreadPoolExecutor(1,
                BasicThreadFactory.Builder().namingPattern("schedule-pool-%d").daemon(true).build())
        executorService.scheduleAtFixedRate({
            logger.info("当前下载百分比为：" + downloadManager.getReadableDownloadedPercent(missionId))

            series.add(second, downloadManager.totalSpeed)
            second = second.next() as Second
        }, 0, 1, TimeUnit.SECONDS)
    }
}

var logger: Logger = LoggerFactory.getLogger(DownloaderSwing::class.java)
fun main() {

    try {
        UIManager.setLookAndFeel("com.sun.java.swing.plaf.windows.WindowsLookAndFeel")
    } catch (e: ClassNotFoundException) {
        logger.error(e.message, e)
    } catch (e: InstantiationException) {
        logger.error(e.message, e)
    } catch (e: IllegalAccessException) {
        logger.error(e.message, e)
    } catch (e: UnsupportedLookAndFeelException) {
        logger.error(e.message, e)
    }

    DownloaderSwing()
}

